trustfall_core/src/interpreter/hints/vertex_info.rs

```diff
@@ -132,6 +146,58 @@ impl<T: InternalVertexInfo + super::sealed::__Sealed> VertexInfo for T {
         }
     }
 
+    /// required_properties returns an iterator over all properties needed
+    /// for this vertex.
+    fn required_properties(&self) -> Box<dyn Iterator<Item = RequiredProperty>> {
+        let current_component = self.current_component();
+
+        let current_vertex = self.current_vertex();
+
+        // we need all output properties
+        let mut properties: Vec<RequiredProperty> = current_component
+            .outputs
+            .values()
+            .filter(|c| c.vertex_id == current_vertex.vid)
+            .map(|c| RequiredProperty::new(c.field_name.clone()))
+            .collect::<Vec<RequiredProperty>>();
+
+        // properties.extend(
+        //     current_component
+        //         .vertices
+        //         .values()
+        //         .filter(|v| v.vid != current_vertex.vid)
+        //         .map(|v| RequiredProperty::new(v.type_name.clone()))
+        //         .collect::<Vec<RequiredProperty>>(),
+        // );
+
+        properties.extend(
+            current_vertex
+                .filters
+                .iter()
+                .map(|f| RequiredProperty::new(f.left().field_name.clone()))
+                .collect::<Vec<RequiredProperty>>(),
+        );
+
+        properties.extend(current_component.vertices.values().flat_map(|v| {
+            v.filters
+                .iter()
+                .filter(|f| match f.right() {
+                    Some(Argument::Tag(FieldRef::ContextField(ctx))) => {
+                        current_vertex.vid == ctx.vertex_id
+                    }
+                    Some(Argument::Tag(FieldRef::FoldSpecificField(fsf))) => {
+                        current_vertex.vid == fsf.fold_root_vid
+                    }
+                    _ => false,
+                })
+                .map(|f| f.right().unwrap().as_tag().unwrap().field_name().into())
```

Comment:
TODO: improve this line

Reply:
Using `.filter_map()` to combine the pattern-match with the mapping might be the way to go here.

Reply:
wow, this is amazing! Thank you!